src/mpi-base.jl

```diff
@@ -5,11 +5,73 @@ typealias MPIDatatype Union{Char,
 typealias MPIBuffertype{T} Union{Ptr{T}, Array{T}, Ref{T}}
 
 # Define a function mpitype(T) that returns the MPI datatype code for
-# a given type T. The dictonary is defined in __init__ so the module
-# can be precompiled
+# a given type T. In the case the the type does not exist, it is created and
+# then returned. The dictonary is defined in __init__ so the module can be
+# precompiled
+
+# accessor and creation function for getting MPI datatypes
+function mpitype{T}(::Type{T})
+
+    if haskey(mpitype_dict, T)  # if the datatype already exists
```

Comment:
I'd had done this split because @JaredCrean2 wanted to ability to add types that were not `bitstype`

It would be odd to pass structs which have pointers (who's data is not passed), but I couldn't see a reason for excluding it if this is what the user really wanted. The default behavior of throwing an error not-`bitstype` seems very reasonable to me.